malleefowl/persist.py

```diff
@@ -8,38 +9,67 @@
 LOGGER = logging.getLogger("PYWPS")
 
 
-def resolve(location, place_holders, f):
+def resolve(location, f, defaults=None):
+    place_holders = re.findall('\{(.*?)\}', location)
+    if not place_holders:
+        return location
+
+    if defaults:
+        facet_values = copy.deepcopy(defaults)
+    else:
+        facet_values = {}
+
     try:
-        nc = netCDF4.Dataset(f, 'r')
+        nc = netCDF4.Dataset(f, 'a')
     except:
-        nc = None
+        return location.format(**facet_values)
 
-    facet_values = {}
-    for facet in place_holders:
-        if hasattr(nc, facet):
-            value = getattr(nc, facet)
-        elif hasattr(nc, facet + '_id'):
-            value = getattr(nc, facet + '_id')
+    if 'time' in nc.variables:
+        nctime = nc.variables['time']
+        if 'calendar' in nctime.ncattrs():
+            calendar = nctime.getncattr('calendar')
         else:
-            value = 'unknown'
-        facet_values[facet] = value
+            calendar = 'gregorian'
+        datetime_i = netCDF4.num2date(nctime[0], nctime.units, calendar)
+        facet_values['initial_year'] = str(datetime_i.year)
+        facet_values['initial_month'] = str(datetime_i.month).zfill(2)
+        facet_values['initial_day'] = str(datetime_i.day).zfill(2)
+        datetime_f = netCDF4.num2date(nctime[-1], nctime.units, calendar)
+        facet_values['final_year'] = str(datetime_f.year)
+        facet_values['final_month'] = str(datetime_f.month).zfill(2)
+        facet_values['final_day'] = str(datetime_f.day).zfill(2)
+        # Could also add support for hour, minute, second, etc.
+
+        if facet_values['initial_year'] == facet_values['final_year']:
+            facet_values['year'] = facet_values['initial_year']
+            if facet_values['initial_month'] == facet_values['final_month']:
+                facet_values['month'] = facet_values['initial_month']
+                if facet_values['initial_day'] == facet_values['final_day']:
+                    facet_values['day'] = facet_values['initial_day']
+
+    for facet in place_holders:
+        if facet in nc.ncattrs():
+            facet_values[facet] = nc.getncattr(facet)
+        elif "{0}_id".format(facet) in nc.ncattrs():
+            facet_values[facet] = nc.getncattr("{0}_id".format(facet))
+        elif facet in defaults:
```

Comment:
no, at that point I don't want to add the time components to the netcdf global attributes, this is redundant, I just want to add the stuff in defaults that is not found in the ncattrs